raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -871,15 +875,32 @@ private FetchResponseData buildFetchResponse(
                 .setLeaderEpoch(quorum.epoch())
                 .setLeaderId(quorum.leaderIdOrNil());
 
-            divergingEpoch.ifPresent(partitionData::setDivergingEpoch);
+            switch (validatedOffsetAndEpoch.type()) {
+                case DIVERGING:
+                    partitionData.divergingEpoch()
+                        .setEpoch(validatedOffsetAndEpoch.offsetAndEpoch().epoch)
+                        .setEndOffset(validatedOffsetAndEpoch.offsetAndEpoch().offset);
+                    break;
+                case SNAPSHOT:
+                    partitionData.snapshotId()
+                        .setEpoch(validatedOffsetAndEpoch.offsetAndEpoch().epoch)
+                        .setEndOffset(validatedOffsetAndEpoch.offsetAndEpoch().offset);
+                    break;
+                default:
+            }
         });
     }
 
     private FetchResponseData buildEmptyFetchResponse(
         Errors error,
         Optional<LogOffsetMetadata> highWatermark
     ) {
-        return buildFetchResponse(error, MemoryRecords.EMPTY, Optional.empty(), highWatermark);
+        return buildFetchResponse(
+            error,
+            MemoryRecords.EMPTY,
+            ValidatedFetchOffsetAndEpoch.valid(new OffsetAndEpoch(-1, -1)),
```

Comment:
Not sure it's worth creating another type, but it is a little surprising to see `valid(new OffsetAndEpoch(-1, -1))`.